ui/src/tempVars/components/EditTemplateVariable.tsx

```diff
@@ -0,0 +1,271 @@
+                value={nextTemplate.tempVar}
+                onChange={this.handleChangeName}
+                onKeyUp={this.handleNameKeyUp}
+                onBlur={this.formatName}
+              />
+            </div>
+            <div className="form-group template-type">
+              <label>Type</label>
+              <Dropdown
+                items={TEMPLATE_TYPES_LIST}
+                onChoose={this.handleChooseType}
+                selected={this.dropDownSelection}
+                buttonSize=""
+              />
+            </div>
+          </div>
+          <div className="edit-temp-var--body-row">
+            <TemplateBuilder
+              template={nextTemplate}
+              source={source}
+              onUpdateTemplate={this.handleUpdateTemplate}
+            />
+          </div>
+          <ConfirmButton
+            text={this.isDeleting ? 'Deleting...' : 'Delete'}
+            confirmAction={this.handleDelete}
+            type="btn-danger"
+            size="btn-xs"
+            customClass="delete"
+            disabled={isNew || this.isDeleting}
+          />
+        </div>
+      </div>
+    )
+  }
+
+  private get templateBuilder(): ComponentClass<TemplateBuilderProps> {
+    const {
+      nextTemplate: {type},
+    } = this.state
+
+    const component = TEMPLATE_BUILDERS[type]
+
+    if (!component) {
+      throw new Error(`Could not find template builder for type "${type}"`)
+    }
+
+    return component
+  }
+
+  private handleUpdateTemplate = (nextTemplate: Template): void => {
+    this.setState({nextTemplate})
+  }
+
+  private handleChooseType = ({type}) => {
+    const {
+      nextTemplate: {id},
+    } = this.state
+
+    const nextNextTemplate = {...DEFAULT_TEMPLATES[type](), id}
+
+    this.setState({nextTemplate: nextNextTemplate})
+  }
+
+  private handleChangeName = (e: ChangeEvent<HTMLInputElement>): void => {
+    const {value} = e.target
+    const {nextTemplate} = this.state
+
+    this.setState({
+      nextTemplate: {
+        ...nextTemplate,
+        tempVar: value,
+      },
+    })
+  }
+
+  private formatName = () => {
+    const {nextTemplate} = this.state
+    const tempVar = formatName(nextTemplate.tempVar)
+
+    this.setState({nextTemplate: {...nextTemplate, tempVar}})
+  }
+
+  private handleSave = async (): Promise<any> => {
+    if (!this.canSave) {
+      return
+    }
+
+    const {onUpdate, onCreate} = this.props
+    const {nextTemplate, isNew} = this.state
+
+    nextTemplate.tempVar = formatName(nextTemplate.tempVar)
+
+    this.setState({savingStatus: RemoteDataState.Loading})
+
+    try {
+      if (isNew) {
+        await onCreate(nextTemplate)
+      } else {
+        await onUpdate(nextTemplate)
+      }
+    } catch (error) {
+      console.error(error)
+    }
+  }
+
+  private handleNameKeyUp = (e: KeyboardEvent<HTMLInputElement>) => {
+    if (e.key === 'Enter') {
+      this.handleSave()
+    }
+  }
+
+  private get isSaving(): boolean {
```

Comment:
👍 thanks. some further thoughts per our earlier discussion: i'm think this makes sense for simple state stuff, whereas more computation heavy class instance methods i'm thinking could simply be invoked in sequential code and assigned to a local var, rather than using a getter which would obfuscate that computational slowness.